chain-tx-enclave/mls/src/keypackage.rs

```diff
@@ -0,0 +1,315 @@
+    pub fn sign(&self, priv_key: &[u8]) -> Vec<u8> {
+        let msg = self.get_encoding();
+        let key_pair =
+            EcdsaKeyPair::from_pkcs8(&signature::ECDSA_P256_SHA256_ASN1_SIGNING, priv_key).unwrap();
+        let system_random = SystemRandom::new();
+        let signature = key_pair.sign(&system_random, &msg).unwrap();
+        signature.as_ref().to_vec()
+    }
+
+    fn find_extension<T: MLSExtension>(&self) -> Result<T, FindExtensionError> {
+        let data = self
+            .extensions
+            .iter()
+            .filter_map(|ext::ExtensionEntry { etype, data }| {
+                if *etype == T::EXTENSION_TYPE {
+                    Some(data)
+                } else {
+                    None
+                }
+            })
+            .next()
+            .ok_or(FindExtensionError(T::EXTENSION_TYPE, "extension not found"))?;
+        // FIXME check remaining data, need support from Codec trait
+        <T>::read_bytes(data).ok_or(FindExtensionError(
+            T::EXTENSION_TYPE,
+            "extension decoding fails",
+        ))
+    }
+
+    /// Verify key package payload
+    pub fn verify(
+        &self,
+        ra_config: EnclaveCertVerifierConfig,
+        now: Timespec,
+    ) -> Result<Vec<u8>, KeyPackageVerifyError> {
+        if self.cipher_suite != MLS10_128_DHKEMP256_AES128GCM_SHA256_P256 {
+            return Err(KeyPackageVerifyError::UnsupportedCipherSuite(
+                self.cipher_suite,
+            ));
+        }
+
+        // Check for required extensions
+        let versions = self.find_extension::<ext::SupportedVersionsExt>()?;
+        if !versions.0.contains(&PROTOCOL_VERSION_MLS10) {
+            return Err(KeyPackageVerifyError::InvalidSupportedVersions);
+        }
+
+        let ciphersuites = self.find_extension::<ext::SupportedCipherSuitesExt>()?;
+        if !ciphersuites
+            .0
+            .contains(&MLS10_128_DHKEMP256_AES128GCM_SHA256_P256)
+        {
+            return Err(KeyPackageVerifyError::InvalidSupportedCipherSuites);
+        }
+
+        let expire = self.find_extension::<ext::ExpirationExt>()?.0;
+        println!("verify expire: {} {}", now, expire);
+        if now > expire {
+            return Err(KeyPackageVerifyError::Expired(expire));
+        }
+
+        let x509 = self
+            .credential
+            .x509()
+            .ok_or(KeyPackageVerifyError::InvalidCredential)?;
+
+        let verifier = EnclaveCertVerifier::new(ra_config)
+            .map_err(KeyPackageVerifyError::VerifierInitError)?;
+        verifier
+            .verify_cert(x509)
+            .map_err(KeyPackageVerifyError::CertificateVerifyError)
+    }
+}
+
+/// Key package, only send `(payload, signature)` to other nodes.
+pub struct KeyPackage {
+    pub payload: KeyPackagePayload,
+    pub signature: Vec<u8>,
+    /// only have the secret key of our own key package
+    pub priv_key: Option<<Kex as KeyExchange>::PrivateKey>,
+}
+
```

Comment:
the golang version has it like that, but it has some extra annotation on the private key.
perhaps a bit more sane would be to have in a separate struct -- e.g. `KeyPackageWrapper` that contains keypackage and corresponding private key to make it explicit.

(There's a similar problem with `Certificate` where here, it's not only the rustls cert, but a wrapper that contains the private key as well... again, may be good to rename, so one doesn't accidentally bundle `Certificate` instead of `rustls::Certificate`)

Reply:
> It is dangerous to store the private key in a format that's meant as publicly exchangeable -- `KeyPackage` is meant to be stored in the ratchet tree

When serializing with rustls's Codec, the private key is skipped, and when deserializing from the network, the private key would be a `None`. Maybe a standalone data type? like `OwnedKeyPackage`?

Reply:
Yes, a standalone data type makes it more explicit

Reply:
Done, `OwnedKeyPackage`.